README.md
# concourse-ci-connector README

This is the README for your extension "concourse-ci-connector". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "concourse-ci-connector" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
/*
 * File: extension.ts
 * Created on: Wednesday, 2023-07-26 @ 10:21:21
 * Author: HackXIt (<hackxit@gmail.com>)
 * -----
 * Last Modified: Thursday, 2023-07-27 @ 15:24:35
 * Modified By:  HackXIt (<hackxit@gmail.com>) @ SE6802S
 * -----
 */

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { ConcourseManager } from './concourseManager';
import { ConcourseExplorerProvider } from './concourseExplorer';

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "concourse-ci-connector" is now active!');

    let concourseManager = ConcourseManager.initialize(context.extensionPath);

    context.subscriptions.push(vscode.commands.registerCommand('concourse-ci-connector.configureHost', () => {
        return concourseManager.configureHost();
    }));
	context.subscriptions.push(vscode.commands.registerCommand('concourse-ci-connector.downloadFlyCli', async () => {
        if(!concourseManager.isHostConfigured()) {
            return vscode.window.showErrorMessage('Concourse host not configured. Please configure the host first using command "Configure host".');
        }
        try {
            const version = await concourseManager.downloadFlyCli();
            await vscode.window.showInformationMessage('Successfully downloaded fly CLI version ' + version);
        } catch(err) {
            if(err instanceof Error) {
                await vscode.window.showErrorMessage(`Error during command: ${err.message}`);
            } else {
                await vscode.window.showErrorMessage(`Unknown error during command: ${err}`);
            }
        }
    }));
    context.subscriptions.push(vscode.commands.registerCommand('concourse-ci-connector.helloWorld', () => {
		// The code you place here will be executed every time your command is executed
		// Display a message box to the user
		vscode.window.showInformationMessage('Hello World from Concourse CI connector!');
	}));

    const workspaceFolders = vscode.workspace.workspaceFolders;

    if (workspaceFolders && workspaceFolders.length > 0) {
        const concourseExplorerProvider = new ConcourseExplorerProvider(workspaceFolders);
        vscode.window.registerTreeDataProvider('concourse-ci-connector', concourseExplorerProvider);
    } else {
        // Handle case when there are no workspace folders
        const concourseExplorerProvider = new ConcourseExplorerProvider([]);
        vscode.window.registerTreeDataProvider('concourse-ci-connector', concourseExplorerProvider);
    }
}

// This method is called when your extension is deactivated
export function deactivate() {}

src/concourseExplorer.ts
/*
 * File: concourseExplorer.ts
 * Created on: Thursday, 2023-07-27 @ 14:51:18
 * Author: HackXIt (<hackxit@gmail.com>)
 * -----
 * Last Modified: Thursday, 2023-07-27 @ 15:21:34
 * Modified By:  HackXIt (<hackxit@gmail.com>) @ SE6802S
 * -----
 */

import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';

export class FileExplorer {
  constructor(private workspaceRoot: string) {
    vscode.window.createTreeView('fileExplorer', { treeDataProvider: new FileExplorerProvider(workspaceRoot) });
  }
}

class File extends vscode.TreeItem {
  constructor(public readonly path: string, public readonly collapsibleState: vscode.TreeItemCollapsibleState) {
    super(path, collapsibleState);
  }
}

class FileExplorerProvider implements vscode.TreeDataProvider<File> {
  constructor(private workspaceRoot: string) {}

  getTreeItem(element: File): vscode.TreeItem {
    return element;
  }

  getChildren(element?: File): Thenable<File[]> {
    if (!this.workspaceRoot) {
      vscode.window.showInformationMessage('No file in empty workspace');
      return Promise.resolve([]);
    }

    if (element) {
      const files = fs.readdirSync(element.path);
      return Promise.resolve(files.map(file => new File(path.join(element.path, file), fs.statSync(path.join(element.path, file)).isDirectory() ? vscode.TreeItemCollapsibleState.Collapsed : vscode.TreeItemCollapsibleState.None)));
    }
    else {
      const files = fs.readdirSync(this.workspaceRoot);
      return Promise.resolve(files.map(file => new File(path.join(this.workspaceRoot, file), fs.statSync(path.join(this.workspaceRoot, file)).isDirectory() ? vscode.TreeItemCollapsibleState.Collapsed : vscode.TreeItemCollapsibleState.None)));
    }
  }
}

export class ConcourseExplorerProvider implements vscode.TreeDataProvider<ConcourseItem> {

    private _onDidChangeTreeData: vscode.EventEmitter<ConcourseItem | undefined> = new vscode.EventEmitter<ConcourseItem | undefined>();
    readonly onDidChangeTreeData: vscode.Event<ConcourseItem | undefined> = this._onDidChangeTreeData.event;

    constructor(private workspaceFolders: readonly vscode.WorkspaceFolder[] | null) {
        this.workspaceFolders = workspaceFolders;
    }

    // Overload signatures:
    refresh(): void;
    refresh(item: ConcourseItem): void;

    // Overload implementation:
    refresh(item?: ConcourseItem): void {
        this._onDidChangeTreeData.fire(item);
    }

    getTreeItem(element: ConcourseItem): vscode.TreeItem {
        return element;
    }

    getChildren(element?: ConcourseItem): Thenable<ConcourseItem[]> {
        if (element) {
            return Promise.resolve([]);
        } else {
            return Promise.resolve(this.getRootConcourseItems());
        }
    }

    /**
     * Generate TreeItem nodes for the root of the explorer view.
     * Here you could add your real implementation fetching the 
     * pipelines or any other information from your concourse instance.
     */
    private getRootConcourseItems(): ConcourseItem[] {
        const item1 = new ConcourseItem('Pipelines', vscode.TreeItemCollapsibleState.Collapsed);
        const item2 = new ConcourseItem('Jobs', vscode.TreeItemCollapsibleState.Collapsed);
        return [item1, item2];
    }
}

class ConcourseItem extends vscode.TreeItem {

    constructor(
        public readonly label: string,
        public readonly collapsibleState: vscode.TreeItemCollapsibleState,
    ) {
        super(label, collapsibleState);
    }

    contextValue = 'concourseItem';

}


src/concourseManager.ts
/*
 * File: concourseManager.ts
 * Created on: Wednesday, 2023-07-26 @ 11:47:57
 * Author: HackXIt (<hackxit@gmail.com>)
 * -----
 * Last Modified: Wednesday, 2023-07-26 @ 14:57:05
 * Modified By:  HackXIt (<hackxit@gmail.com>) @ SE6802S
 * -----
 */

import * as https from 'https';
import * as fs from 'fs';
import * as os from 'os';
import * as vscode from 'vscode';

export class ConcourseManager {
    private static instance: ConcourseManager;
    // Concourse host for this instance
    private host: string;
    // Path for the Fly CLI
    private cliPath: string;
    private flyPath: string;
    private flyConfigured: boolean;
    private hostConfigured: boolean;

    // Private constructor so it can't be directly instantiated
    private constructor(cliPath: string) {
        this.host = '';
        this.flyPath = '';
        this.flyConfigured = false;
        this.hostConfigured = false;
        this.cliPath = cliPath;
    }

    private static async executeFly(flyPath: string, args: string[]): Promise<string> {
        return new Promise((resolve, reject) => {
            const fly = require('child_process').spawn(flyPath, args, { shell: true });
            let output = '';
            fly.stdout.on('data', (data: string) => {
                output += data;
            });
            fly.stderr.on('data', (data: string) => {
                output += data;
            });
            fly.on('close', (code: number) => {
                if(code !== 0) {
                    reject(new Error('Failed to execute fly CLI: ' + output));
                    return;
                }
                resolve(output);
            });
        });
    }

    // Method to get an instance of ConcourseManager
    public static initialize(flyPath: string): ConcourseManager {
        if (!ConcourseManager.instance) {
            ConcourseManager.instance = new ConcourseManager(flyPath);
        } else {
            ConcourseManager.instance.cliPath = flyPath;
            let host = vscode.workspace.getConfiguration('concourse-ci-connector').get<string>('host');
            if(host !== undefined) {
                ConcourseManager.instance.host = host;
                ConcourseManager.instance.hostConfigured = true;
            } else {
                ConcourseManager.instance.host = '';
                ConcourseManager.instance.hostConfigured = false;
            }
        }
        return ConcourseManager.instance;
    }

    public static getInstance(): ConcourseManager {
        if(!ConcourseManager.instance) {
            throw new Error('ConcourseManager not initialized.');
        }
        return ConcourseManager.instance;
    }

    public isHostConfigured(): boolean {
        return this.hostConfigured; 
    }

    public async downloadFlyCli(): Promise<string> {
        return new Promise((resolve, reject) => {
            const platform = os.platform();
            const arch = os.arch();
            let platformString = '';
            let archString = '';
            let filePath = this.cliPath + '/fly';
            if (platform === 'win32') {
                filePath += '.exe';
                platformString = 'windows';
            } else {
                platformString = platform;
            }
            if (arch === 'x64') {
                archString = 'amd64';
            } else {
                archString = arch;
            }
            const url = `https://${this.host}/api/v1/cli?arch=${archString}&platform=${platformString}`;
            console.debug('Downloading fly CLI from ' + url);

            const file = fs.createWriteStream(filePath);
            https.get(url, function(response) {
                if(response.statusCode !== 200) {
                    reject(new Error('Failed to download fly CLI: ' + response.statusMessage + ' (HTTP Status Code: ' + response.statusCode + ')'));
                    return;
                }
                response.pipe(file);
                file.on('finish', function() {
                    file.close();
                    // Make the file executable on unix systems
                    if(platform !== 'win32') {
                        fs.chmodSync(filePath, '755');
                    }
                    ConcourseManager.instance.flyPath = filePath;
                    ConcourseManager.instance.flyConfigured = true;
                    // Execute fly --version to verify the downloaded binary
                    ConcourseManager.executeFly(filePath, ['--version'])
                        .then((output) => resolve(output))
                        .catch((error) => reject(error));
                });
            }).on('error', function(err) {
                fs.unlinkSync(filePath);
                reject(err);
            });
        });
    }

    public async configureHost(): Promise<void> {
        let host = await ConcourseManager.askForHost();
        if(host !== undefined) {
            this.host = host;
            this.hostConfigured = true;
            vscode.workspace.getConfiguration('concourse-ci-connector').update('host', host, true);
            await vscode.window.showInformationMessage('Host configured: ' + host);
            console.log('Host configured: ' + host);
        } else {
            console.log('Host configuration cancelled.');
        }
    }

    public static async askForHost(): Promise<string|undefined> {
        return vscode.window.showInputBox(
            {
                prompt: 'Enter the host URL for your Concourse CI instance.',
                placeHolder: 'concourse.example.com'
            }
        );
    }
}